src/V3DfgPasses.cpp

```diff
@@ -275,6 +244,120 @@ void V3DfgPasses::removeUnused(DfgGraph& dfg) {
+    // if an element is in the list by checking for a non-zero next pointer, and easy
+    // prefetching without conditionals. The address of the graph is a good sentinel as it is a
+    // valid memory address, and we can easily check for the end of the list.
+    DfgVertex* const sentinelp = reinterpret_cast<DfgVertex*>(&dfg);
+    DfgVertex* workListp = sentinelp;
+
+    // Add all variables to the initial work list
+    for (DfgVertexVar *vtxp = dfg.varVerticesBeginp(), *nextp; vtxp; vtxp = nextp) {
+        nextp = vtxp->verticesNext();
+        if (VL_LIKELY(nextp)) VL_PREFETCH_RW(nextp);
+        vtxp->setUser<DfgVertex*>(workListp);
+        workListp = vtxp;
+    }
+
+    const auto addToWorkList = [&](DfgVertex& vtx) {
+        // If already in work list then nothing to do
+        DfgVertex*& nextInWorklistp = vtx.user<DfgVertex*>();
+        if (nextInWorklistp) return;
+        // Actually add to work list.
+        nextInWorklistp = workListp;
+        workListp = &vtx;
+    };
+
+    // Variable replacements to apply in the module
+    std::unordered_map<AstVar*, AstVar*> replacements;
+
+    // Process the work list
+    while (workListp != sentinelp) {
+        // Pick up the head of the work list
+        DfgVertex* const vtxp = workListp;
+        // Detach the head
+        workListp = vtxp->getUser<DfgVertex*>();
+        // Prefetch next item
+        VL_PREFETCH_RW(workListp);
+
+        // Remove unused non-variable vertices
+        if (!vtxp->is<DfgVertexVar>() && !vtxp->hasSinks()) {
+            // Add sources of removed vertex to work list
+            vtxp->forEachSource(addToWorkList);
+            // Remove the unused vertex
+            vtxp->unlinkDelete(dfg);
+        }
+
+        // We can only eliminate DfgVarPacked vertices at the moment
+        DfgVarPacked* const varp = vtxp->cast<DfgVarPacked>();
+        if (!varp) continue;
+
+        // Can't remove if it has external drivers
+        if (!varp->isDrivenFullyByDfg()) continue;
+
+        // Can't remove if must be kept (including external, non module references)
+        if (varp->keep()) continue;
+
+        // Can't remove if referenced in other DFGs of the same module (otherwise might rm twice)
+        if (varp->hasDfgRefs()) continue;
+
+        // If it has multiple sinks, it can't be eliminated
+        if (varp->hasMultipleSinks()) continue;
+
+        if (!varp->hasModRefs()) {
+            // If it is only referenced in this DFG, it can be removed
+            ++ctx.m_varsRemoved;
+            varp->replaceWith(varp->source(0));
+            varp->varp()->unlinkFrBack()->deleteTree();
+        } else if (DfgVarPacked* const canonp = varp->source(0)->cast<DfgVarPacked>()) {
+            // If it's driven from another canonical variable, it can be replaced by that.
+            // However, we don't want to propagate SystemC variables into the design
+            if (canonp->varp()->isSc()) continue;
```

Comment:
This really is just for SystemC (we don't want to replace references to non-SC vars with SC vars/we want SC vars to be used only on the top-level IO). It's ok to replace a reference to a non-public signal with a reference to an equivalent public signal. 

Public signals are otherwise preserved by checking `keep()` above.